### ISSUE TITLE
Client side data architecture

### ISSUE THREAD

**Issue by kirlat** [alpheios-project/documentation#17]
CEDICT requires downloading and storing large amounts of lexical data on the client side. We're also using significant amount of lexical data for inflection tables. 

It would be good to discuss and decide on what would be the best way to handle such data. We can use it as a unified architecture for cases described above and for any similar future uses that may arise.

The most important parts of the architecture is about in what format data would be, from where and how would an application get the data, and how would this lexical data be stored and used.

On my opinion, the best format for the data is JSON. It is a de-facto standard for JS and other languages. It is not limited to tabular data as CSV is; it can describe complex data structures as well. JSON is easily readable by humans. There are libraries implementing its support in different programming languages. There are also many editor-like tools that work with JSON. Comparing with XML, JSON is more compact and less trouble to understand.

The drawbacks of JSON is that it is hard to create and edit in large volumes manually. The other disadvantage is that JSON can still be more compact than what it is.

To mitigate the first issue we, for tabular data, can create a script to convert a source from CSV (that can be easily edited in spreadsheet applications) to JSON. We can also offer the use specialized JSON editing tools for more complex cases.

For the second issue we can use compression libraries such as [LZ-String](https://github.com/pieroxy/lz-string/) or server-based compression.

The best way to store large amounts of data permanently is IndexedDB. To speed up operations some objects can be read from IndexedDB to RAM. For example, with CEDICT we can keep the main body of a dictionary in IndexedDB but use RAM for indexes. This way we'll be able to find the right key fast and issue a query to retrieve a record. That strategy may vary from device to device: on devices with smaller amount of RAM we can store everything in IndexedDB. On better equipped machines we can store everything in RAM. We can also give user an ability to choose wither to have data in RAM or in IndexedDB or in both. The flexibility to work with both RAM and IndexedDB stored data transparently shall be a part of our architecture.

How would the data get into IndexedDB store initially? We can either download it from the server upon the first initialization of an app or we can bundle it with the application. We can combine both approaches as well: some data can be downloaded dynamically and some can be bundled statically.

We might need to provide different types of bundles (differ in what amount of data is downloaded and bundled) for different users and purposes. It would be good to have settings for that as a build time config; with it, we can rebuild into different types of bundles without changing a line in the source code.

To avoid re-downloading significant amounts of data upon a connection interruption we can split large data files into chunks and load them individually. Setting a chunk size threshold to 1MB (uncompressed size) would, on my opinion, make a good compromise between the risks of re-downloading and the hassle of splitting files into to many fragments. So, if we accept this, everything over 1MB must be split.

Users experiencing most problems with downloading would be the ones connecting over mobile networks, most likely. Here is FYI a [mobile network speed guide](Mobile speed guide: https://www.tigermobiles.com/faq/mobile-download-speed-guide/) describing real life download speeds for mobile users. There is also a [mobile speed calculator](https://www.broadbandsolutions.com.au/business-centre/viewpoint/understanding-connection-speeds-megabytes-megabits) to check how long will it take to load a single chunk of data.

The simplest way to expose data for downloading is to put it to an HTTPS server as JSON files. If server will support gzip or other type of compression (and most do), that will solve the issue of minimizing the incoming traffic. We can also use CDN to make connection more efficient.

To keep track of all the resources we can have a central TOC file in JSON format that will provide all types of lexical resource available for downloading. It will have the names and the locations of the files as well as chunks. An app will load this file first and then will use information from it to download pieces of data that it does not have.

All data files shall be versioned so that we'll be able to perform data upgrades without disrupting functionality of users with previous versions of an app.  We can have version number in both file names and names of IndexedDB object stores. We can also have it inside an in-memory JS object that will be representing lexical data.

**Comment by balmas** [comment]
another consideration for the design: we have multiple requests, particularly from users of the embedded library, to be able to user their own data files, e.g. for inflection tables and lexical resources.   Whatever design we come up with should be able to accommodate that easily

**Comment by balmas** [comment]
per discussion at today's check-in, let's proceed to experiment with and explore this architecture.

**Comment by irina060981** [comment]
From my point of view division to 1Mb portions is not very useful, because for todays internet resources standards - 1Mb is a very small amount.
For example:
    http://alpheios.net/ - 4.3Mb
    https://texts.alpheios.net/- 3.2Mb
    https://www.loebclassics.com/ - 3.8Mb

In our area - on Far East if we have Internet connection (we have places without mobile internet connections at all), than we have 3G/4G/LTE.

For example, for all chinese json files you need nearly 25Mb (9.1, 7, 4.6, 4.6, 1) = 28 chunks - that should be organized, controlled and deleted if the upload is failed.

For Chinese (Inflections, Grammar) we are going to make the user decide - does he/she want to upload all files to use chinese, so he/she would upload it in the area with Internet connection that is appropriate for this task, I believe.

That's why I think that there is no some strict need to have so small chunks.

Other points are well!

**Comment by kirlat** [comment]
> From my point of view division to 1Mb portions is not very useful, because for todays internet resources standards - 1Mb is a very small amount.

I think that's a very good point. I did not realize how small 1MB is by today's standards and how many chunks would we need to have to split all of the Chinese resources.

I'll try to experiment with compression of the data to see what ratios could we get and to what limit would we be able to rise the uncompressed file's size limit up to.

**Comment by balmas** [comment]
I would also like to understand a little bit more how this approach overlaps with or could use PWA strategies.  @kirlat can you expand on that a bit? Thanks!

**Comment by balmas** [comment]
What you're suggesting makes sense when comparing the use of the Cache API vs IndexedDb API. But wouldn't we want to use service workers to manage syncing and versioning?

**Comment by kirlat** [comment]
> But wouldn't we want to use service workers to manage syncing and versioning?

Background synchronization and upgrades are important features of the PWA that we would be not having without it, I agree. They're probably not so important for us now, but we could definitely benefit from it in the future.

**Comment by balmas** [comment]
I think that sounds good!

**Comment by kirlat** [comment]
There are several ways we can represent dictionary data in a file.

It can be simply a serialized array of array values:
```
[
    66621,
    "熱核聚變反應",
    "热核聚变反应",
    "re4 he2 ju4 bian4 fan3 ying4",
    "thermonuclear fusion reaction"
  ]
```

We can store it as a an array of objects with shorter filed names:
```
  {
    "idx": 66621,
    "tradHW": "熱核聚變反應",
    "simplHW": "热核聚变反应",
    "pinyin": "re4 he2 ju4 bian4 fan3 ying4",
    "def": "thermonuclear fusion reaction"
  }
```

Or it can be an array of objects with longer field names:
```
  {
    "index": 66621,
    "traditionalHeadword": "熱核聚變反應",
    "simplifiedHeadword": "热核聚变反应",
    "pinyin": "re4 he2 ju4 bian4 fan3 ying4",
    "definition": "thermonuclear fusion reaction"
  }
```

I was compressing those files with ZIP. On average, compression ratio was 5:1, i.e. the compressed file was approximately 5 times smaller than the original. However, to my surprise, the compression did not factor away repeating elements (i.e. field names) completely and there was a slight difference in size of compressed files across the three. The compressed files with shorter and longer filed names where approximately 5% and 10% larger than the array-based file.

With all this, I'm not sure what the best schema for storing data would be. Array results in a smaller file, but it is harder to read and edit. Honestly, there will probably be a very little chance that anyone would decide to read or edit this file directly: we have a source file (dat in case of CEDICT) that will be the one updated with new data and the abovementioned JSON file will be created by running a conversion script that will create it. The workflow will be as following: `Source file (in any format possible) -> conversion script -> destination JSON file`. The app code will read the the JSON file and convert it to an array of objects and it probably does not matter for the app in what format it will be.

Yet the files with serialized objects looks more understandable and the size difference with the array-based ones is really small.

At the top of a JSON file we can add an object containing some general info about the dictionary: its origin, version, copyrights, etc. An app will read it and display in credits. If we'll use arrays to store the records, we can also add there the names of the array elements as `[`index`, `traditionalHeadword`, `simplifiedHeadword`, `pinyin`, `definition`]. This would be helpful in providing some reference.

With all this it's really hard to decide which format shall we go with. Do you have any opinions on that?

**Comment by kirlat** [comment]
If we have a special variant of a dictionary data then we can specify that after a dictionary name. Let's say there is a CEDICT version that contains simplified Chinese words only. In that case we can name it like `cedict-simpl-v20191231.json`.

We can specify a chunk number as the last segment of the file name: `cedict-v20191231-c001.json`. The numbering shall probably start with 1 (as it matches the natural way of counting things).

With this naming schema the file names will be concise yet provide enough information about content of the file.

File name shall be all lowercase to avoid problems during transfer between case sensitive and insensitive file systems.

**Comment by kirlat** [comment]
> From my point of view division to 1Mb portions is not very useful, because for todays internet resources standards - 1Mb is a very small amount.

What if we lift it up to 10MB uncompressed? Then all CEDICT data will fit into three chunks at most. Other lexical resources will probably be smaller and will not need more than one chunk. Yet the compressed chunk, if base it on the test results with CEDICT, will be around 2MB and will download pretty fast.

Does this sound good?

**Comment by balmas** [comment]
> For versioning, the best approach would be, on my opinion, to go with the date based model: `20191231` comprised of `yyyymmdd`. This format provides a natural sort order. That maybe does not matter for us now, but can be beneficial for file organization on the server. The version is easy to understand: it is a date when the resource was last updated. That is a meaningful info to users and developers.
> 
> With this, the name of a dictionary file can be `dictname-vyyyymmdd.json` (i.e. `cedict-v20191231.json`).
> 

I am fine with using dates for versioning but I think we should allow for finer granularity than 1 day. It's not inconceivable that we might need to release multiple versions in one day (if we release and find a mistake, for example). I don't know if we need a full timestamp -- a single incrementer value would also work, (e.g. dictname-vyyyymmdd-n.json)

**Comment by balmas** [comment]
>I think that in order to simplify file management we probably shall combine as many different objects into one file as possible. JSON allows us to do that: we can create an object that will have other objects as its props. 

this seems fine

**Comment by balmas** [comment]
> > From my point of view division to 1Mb portions is not very useful, because for todays internet resources standards - 1Mb is a very small amount.
> 
> What if we lift it up to 10MB uncompressed? Then all CEDICT data will fit into three chunks at most. Other lexical resources will probably be smaller and will not need more than one chunk. Yet the compressed chunk, if base it on the test results with CEDICT, will be around 2MB and will download pretty fast.
> 
> Does this sound good?

seems ok to me

**Comment by kirlat** [comment]
> I am fine with using dates for versioning but I think we should allow for finer granularity than 1 day. It's not inconceivable that we might need to release multiple versions in one day (if we release and find a mistake, for example). I don't know if we need a full timestamp -- a single incrementer value would also work, (e.g. dictname-vyyyymmdd-n.json)

Oh, I did not realize there could be more than one version released during a day. In the current schema I was thinking of dashes as group separators (dictionary name/variety/version/chunk) so maybe we can use something else inside the version group so it'll be clear it's part of the version? We can maybe use  `r` as an abbreviation for "revision" followed by the revision number? We can not add it for the first revision but add it if it's a consecutive daily revision: the first revision will be `v20191231` but if there will be a second revision this day it will become `v20191231r2` (because there was an r1 already, it just had a revision number omitted). How would that be? Or, as an alternative, we can add letters at the end going through the alphabet as the revision number grows (i.e. a, b, c, etc.).

**Comment by kirlat** [comment]
> the only concern that occurs to me right now is whether or not there are any security implications (take the case for example where we allow 3rd parties to supply dictionary files). I don't know if we want the ability to allow running arbitrary javascript from external sources.

I was thinking about doing static import of those function only and only from the repositories that we control. If someone created a dictionary I'm not sure it's reasonable for us to expect from him/her to write an import function to integrate it into our system. We would have to make changes to our code to support a new dictionary anyway; then it would be a good time to create an import function too.

If the dictionary will be in a public repository that we can probably fork it, add an importer code, and then sync with upstream changes. If the dictionary is not in a repo, we would probably have to create a repository for it and sync changes manually. I don't think we can allow third parties to integrate dictionaries without reviewing them first. With GitHub based dictionaries we can use version tags and pull the new tag only only after it is reviewed and approved. We can even use PRs to review incoming changed in data.

There are several tools that allow editing files directly on GitHub. So if the third party would use such a tool, or if we would provide such a tool using a GitHub API, the changes will go directly into the repository and that will be an ideal situation IMO.

**Comment by balmas** [comment]
>We can maybe use `r` as an abbreviation for "revision" followed by the revision number? We can not add it for the first revision but add it if it's a consecutive daily revision: the first revision will be `v20191231` but if there will be a second revision this day it will become `v20191231r2` (because there was an r1 already, it just had a revision number omitted). How would that be? Or, as an alternative, we can add letters at the end going through the alphabet as the revision number grows (i.e. a, b, c, etc.).

Either of those is fine with me.

**Comment by balmas** [comment]
>..with GitHub based dictionaries..

Do I understand correctly that you are thinking we will pull the files from GitHub directly at runtime?

**Comment by kirlat** [comment]
> Do I understand correctly that you are thinking we will pull the files from GitHub directly at runtime?

Nope, I was not thinking about that (yet). I don't know if it is possible or is it desirable (GitHub is probably neither supports nor is optimized for that kind of activity). But if we can use GitHub as a CDN (which I doubt is possible) then that can save us an extra step.

I was just thinking about storing all dictionaries on GitHub and that it will be beneficial because of how easily we can track changes and other conveniences it provides. We could then pull dictionaries to our servers to make them available to our users.

**Comment by balmas** [comment]
ok. I am still not sure how I feel about having the importer code stored directly in the dictionary. Could we make this optional? Maybe by having a default function that can be used?

**Comment by kirlat** [comment]
> ok. I am still not sure how I feel about having the importer code stored directly in the dictionary. Could we make this optional? Maybe by having a default function that can be used?

I thought of it as of a convenience measure but maybe it's not needed at all. I agree that its probably a better idea from a security point to use a universal importer instead. We maybe can establish some JSONSchema and control if data conforms to it and maybe we can add some more data checks to the function too.

**Comment by kirlat** [comment]
> Either of those is fine with me.

Then maybe we can chose the one with the `r` at the end (i.e. `v20191231r2`)? I think this one is easier to comprehend (at least to me).

**Comment by balmas** [comment]
>Then maybe we can chose the one with the r at the end (i.e. v20191231r2)? I think this one is easier to comprehend (at least to me).

works for me.

**Comment by balmas** [comment]
implemented in release 3.3.0
